### PR TITLE
Remote server plotting

### LINKED ISSUE #28: should automatically save plots that are displayed
The training curve disappears as soon as the script has finished executing.

### LINKED ISSUE #46: Plotting on remote server may return error with a different matplotlib backend
```
if os.name is not "posix":
    if "DISPLAY" not in os.environ:
        matplotlib.use('Agg')
else:
    matplotlib.use('TkAgg')
import matplotlib.pyplot as plt
```

The biggest issue is that if a user forgets the `plot=true` parameter on a device that doesn't have a `DISPLAY` os environment variable, matplotlib will crash. Perhaps if 'plot=true' and there is nowhere to display, raise error quickly to change plot to false?

### LINKED ISSUE #76: Test Visualization

### PR DESCRIPTION
**Pull request by joemehltretter**
Add functionality to check if system display variable is available for use when `save_plot` is true. Also, added functionality to save produced graphic at a given location. Closes #28, closes #46, and closes #76

### CODE REVIEW

File: vulcanai2/plotters/visualization.py [hunk 1]
```diff
@@ -21,8 +21,44 @@
...
+    None
+
+    """
+    if not assert_display_available() and save_path is None:
+        raise RuntimeError("No display environment found. Display environment needed to plot, or set save_path=path/to/dir")
+    else:
+        plot.savefig(path)
 
-def display_record(record=None, load_path=None, interactive=True):
+def display_record(record=None, load_path=None, save_path=None, interactive=True):
```
[the remark this answers is not in the document]
Comment (joemehltretter) [reply vulcanai2/plotters/visualization.py]: @rfratila Remove all usage of `load_path`?

[the remark this answers is not in the document]
Comment (rfratila) [reply vulcanai2/plotters/visualization.py]: yeah there's only a few lines of code in `display_record`

[the remark this answers is not in the document]
Comment (joemehltretter) [reply vulcanai2/plotters/visualization.py]: @rfratila Completed in most recent push

File: vulcanai2/plotters/visualization.py [hunk 2]
```diff
@@ -133,11 +178,19 @@ def display_tsne(input_data, targets, label_map=None):
...
         label_map=label_map,
-        title='t-SNE Visualization')
+        title='t-SNE Visualization',
+        save_path=save_path)
 
 
-def _plot_reduction(x_transform, targets, label_map, title):
+def _plot_reduction(x_transform, targets, label_map, title, save_path=None):
     """Once PCA and t-SNE has been calculated, this is used to plot."""
+    if save_path:
```
[the remark this answers is not in the document]
Comment (joemehltretter) [reply vulcanai2/plotters/visualization.py]: @rfratila Completed in recent push

File: vulcanai2/plotters/visualization.py [hunk 3]
```diff
@@ -21,8 +21,44 @@
...
+        Matplotlib variable with savefig ability.
+    path : string
+        String that designates the path to save the given figure to.
+
+    Returns
+    -------
+    None
+
+    """
+    if not assert_display_available() and save_path is None:
```
[the remark this answers is not in the document]
Comment (joemehltretter) [reply vulcanai2/plotters/visualization.py]: @rfratila Completed in recent push

File: vulcanai2/plotters/visualization.py [hunk 4]
```diff
@@ -21,8 +21,44 @@
...
+    Check that display variable exists
+
+    Returns
+    -------
+    Boolean
+        True if display variable exists
+        False if display variable does not exist
+
+    """
+    if os.environ.get("DISPLAY"):
```
[the remark this answers is not in the document]
Comment (joemehltretter) [reply vulcanai2/plotters/visualization.py]: Completed in recent push

File: vulcanai2/models/basenetwork.py [hunk 5]
```diff
@@ -563,6 +566,11 @@ def fit(self, train_loader, val_loader, epochs,
                 fig_number = plt.gcf().number + 1 if plt.fignum_exists(1) else 1
                 plt.show()
 
+            if save_path:
+                time = str(datetime.now())
```
Comment (rfratila) [inline vulcanai2/models/basenetwork.py]: you could use your new `get_time()` function here. it may introduce circular dependencies so I would just move that function to model utils

Comment (rfratila) [reply vulcanai2/models/basenetwork.py]: oh sorry there should only be one function declaration. if you wanted it to get used across visualization and models, I would recommend it remain in visualization and models can import from the visualization files

File: vulcanai2/plotters/visualization.py [hunk 6]
```diff
@@ -32,10 +32,7 @@ def assert_display_available():
         False if display variable does not exist
 
     """
-    if os.environ.get("DISPLAY"):
-        return True
-    else:
-        return False
+    True if os.environ.get("DISPLAY") else False
```
[the remark this answers is not in the document]
Comment (joemehltretter) [reply vulcanai2/plotters/visualization.py]: @rfratila I added it

File: vulcanai2/models/basenetwork.py (hunk 5 above)
Comment (joemehltretter) [reply vulcanai2/models/basenetwork.py]: @rfratila okay, completed

**Comment by rfratila** [comment]
@joemehltretter this looks good but would you be able to make the `test_display_tsne` use not all the data from sklearn digits? it's taking a good chunk of time to make the t-sne graph. we only need to use like a quarter or less of that entire dataset

File: vulcanai2/plotters/visualization.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rfratila) [reply vulcanai2/plotters/visualization.py]: @joemehltretter my thinking is that someone can specify the `plot` to be `false` but still want to save the figures to disk by specifying the `save_path` arg

File: vulcanai2/plotters/visualization.py [hunk 7]
```diff
@@ -21,17 +20,56 @@
...
+    Check that display variable exists
+
+    Returns
+    -------
+    Boolean
+        True if display variable exists
+        False if display variable does not exist
+
+    """
+    return True if os.environ.get("DISPLAY") else False
```
Comment (nirtiac) [inline vulcanai2/plotters/visualization.py]: Why not just make this into a global variable? Not sure what is more pythonic but it's certainly simpler to go DISPLAY_AVAILABLE = os.environ.get("DISPLAY")

File: vulcanai2/plotters/visualization.py [hunk 8]
```diff
@@ -21,17 +20,56 @@
...
+        Matplotlib variable with savefig ability.
+    path : string
+        String that designates the path to save the given figure to.
+
+    Returns
+    -------
+    None
+
+    """
+    if not assert_display_available() and path is None:
```
Comment (nirtiac) [inline vulcanai2/plotters/visualization.py]: Wait, but what if the path is None but the display is available?

Comment (nirtiac) [reply vulcanai2/plotters/visualization.py]: In general this doesn't seem relevant to a "save_visualization" function. Why is this here?

File: vulcanai2/plotters/visualization.py [hunk 9]
```diff
@@ -21,17 +20,56 @@
...
+        raise RuntimeError("No display environment found. Display environment needed to plot, or set save_path=path/to/dir")
+    else:
+        plot.savefig(path)
+
+def get_time():
+    curr_time = str(datetime.now())
+    curr_time = curr_time.replace(" ", '_')
+    curr_time = curr_time.split('.')[0]
+    return curr_time
```
Comment (nirtiac) [inline vulcanai2/plotters/visualization.py]: This is what we have for saving the models, could we standardize? 
  `self._save_path = "{}_{date:%Y-%m-%d_%H:%M:%S}/".format(
                self.name, date=datetime.now())`

File: vulcanai2/plotters/visualization.py [hunk 10]
```diff
@@ -41,12 +79,6 @@ def display_record(record=None, load_path=None, interactive=True):
 
     """
     title = 'Training curve'
-    if load_path is not None:
-        with open(load_path) as in_file:
-            record = pickle.load(in_file)
-        title = 'Training curve for model: {}'.format(
-            os.path.basename(load_path))
```
Comment (nirtiac) [inline vulcanai2/plotters/visualization.py]: It would be nice to have a separate load function to replace this, for those who wish to view their images that way.

File: vulcanai2/plotters/visualization.py [hunk 11]
```diff
@@ -201,8 +249,13 @@ def display_confusion_matrix(cm, class_list=None):
     divider = make_axes_locatable(ax)
     cax = divider.append_axes("right", size="4%", pad=0.03)
     plt.colorbar(im, cax=cax)
-    plt.show(False)
 
+    if save_path:
+        time = get_time()
+        save_path = save_path + '/confusion_matrix' + '_' + time + '.png'
```
Comment (nirtiac) [inline vulcanai2/plotters/visualization.py]: you could have a "get_save_path" function that replaces the get_time function, in case we or someone decided they wanted to change the standard save path later on

File: vulcanai2/tests/plotters/test_visualization.py [hunk 12]
```diff
@@ -104,3 +113,47 @@ def test_compute_saliency_map_dnn(self, dnn_class):
...
+        test_input_1B = torch.ones([1, *dnn_class.in_dim])
+        sal_map_1B = np.array(compute_saliency_map(
+            dnn_class,
+            test_input_1B, torch.LongTensor([2])))
+        display_saliency_overlay(img, sal_map_1B, shape=(25,25), save_path=curr_path)
+        for file in os.listdir(curr_path):
+            if file.startswith('saliency'):
+                assert True
+                file_path = curr_path + file
+        os.remove(file_path)
```
Comment (nirtiac) [inline vulcanai2/tests/plotters/test_visualization.py]: os.remove should be indented.

File: vulcanai2/tests/plotters/test_visualization.py [hunk 13]
```diff
@@ -104,3 +113,47 @@ def test_compute_saliency_map_dnn(self, dnn_class):
...
+
+    def test_display_saliency_overlay(self, dnn_class):
+        curr_path = str(os.path.dirname(__file__)) + '/'
+        img = np.zeros((25,25))
+        test_input_1B = torch.ones([1, *dnn_class.in_dim])
+        sal_map_1B = np.array(compute_saliency_map(
+            dnn_class,
+            test_input_1B, torch.LongTensor([2])))
+        display_saliency_overlay(img, sal_map_1B, shape=(25,25), save_path=curr_path)
+        for file in os.listdir(curr_path):
```
Comment (nirtiac) [inline vulcanai2/tests/plotters/test_visualization.py]: Unless I'm mistaken, there's no way to fail this loop?

File: vulcanai2/tests/plotters/test_visualization.py [hunk 14]
```diff
@@ -104,3 +113,47 @@ def test_compute_saliency_map_dnn(self, dnn_class):
...
+            if file.startswith('saliency'):
+                assert True
+                file_path = curr_path + file
+        os.remove(file_path)
+
+    def test_display_pca(self, dnn_class):
+        curr_path = str(os.path.dirname(__file__)) + '/'
+        digits = load_digits()
+        dnn_pca = display_pca(digits.data[0:10], digits.target[0:10], save_path=curr_path)
+        for file in os.listdir(curr_path):
```
Comment (nirtiac) [inline vulcanai2/tests/plotters/test_visualization.py]: ditto re no way to fail, and the tests below.

File: vulcanai2/plotters/visualization.py (hunk 8 above)
Comment (joemehltretter) [reply vulcanai2/plotters/visualization.py]: @nirtiac @rfratila I know we had a reason, but I can't seem to remember it. However, looking at it now it doesn't make sense. If a path is specified we need to save the visualization before plotting and we need to confirm a display variable exists before plotting. We can remove that `assert_display_available()` from the `save_visualization` function and put it in an if statement before `plt.show()` for each display function.

File: vulcanai2/plotters/visualization.py (hunk 9 above)
Comment (joemehltretter) [reply vulcanai2/plotters/visualization.py]: @nirtiac You got it!

File: vulcanai2/tests/plotters/test_visualization.py [hunk 15]
```diff
@@ -122,38 +122,58 @@ def test_display_saliency_overlay(self, dnn_class):
...
         digits = load_digits()
         dnn_tsne = display_tsne(digits.data[0:10], digits.target[0:10], save_path=curr_path)
+        file_created = False
         for file in os.listdir(curr_path):
             if file.startswith('t-SNE'):
-                assert True
+                file_created = True
                 file_path = curr_path + file
                 os.remove(file_path)
+        if file_created:
```
Comment (rfratila) [inline vulcanai2/tests/plotters/test_visualization.py]: you could just have `assert file_created` no @joemehltretter ?

**Comment by rfratila** [comment]
@joemehltretter I get the following warning when I run the tests 
```
PendingDeprecationWarning: the matrix subclass is not the recommended way to represent matrices or deal with linear algebra (see https://docs.scipy.org/doc/numpy/user/numpy-for-matlab-users.html). Please adjust your code to use regular ndarray.
  return matrix(data, dtype=dtype, copy=False)
```
would you be able to update?

File: vulcanai2/plotters/visualization.py (hunk 9 above)
Comment (joemehltretter) [reply vulcanai2/plotters/visualization.py]: @nirtiac @rfratila The only issue I see with this is that `self.name` is not being sent to `visualization`.

Comment (nirtiac) [reply vulcanai2/plotters/visualization.py]: You could have name as the parameter of the function, as you do provide a name/label when you call get_time below

Comment (joemehltretter) [reply vulcanai2/plotters/visualization.py]: @nirtiac Okay I see, that's what I just updated it to do. I'll condense to standardize

**Comment by nirtiac** [comment]
> @rfratila I am not getting that `PendingDeprecationWarning` when running tests? Where exactly is it occurring?

@rfratila @joemehltretter I'm guessing this means that

**Comment by nirtiac** [comment]
Reopening

File: vulcanai2/plotters/visualization.py [hunk 16]
```diff
@@ -49,17 +38,24 @@ def save_visualization(plot, path=None):
...
+    path = path + "{}_{date:%Y-%m-%d_%H:%M:%S}".format(vis_type, date=datetime.now()) + '.png'
+    return path
 
 def get_time():
     curr_time = str(datetime.now())
     curr_time = curr_time.replace(" ", '_')
     curr_time = curr_time.split('.')[0]
     return curr_time
 
+def load_vis(load_path=None):
```
Comment (rfratila) [inline vulcanai2/plotters/visualization.py]: @nirtiac @joemehltretter should we pass it a network file path to load and then use basenetwork.load_model to get the record/display it?

Comment (joemehltretter) [reply vulcanai2/plotters/visualization.py]: @rfratila so something along the lines of `model = basenetwork.load_model(model_path)` then send `model.record` to `display_record`?

Comment (rfratila) [reply vulcanai2/plotters/visualization.py]: Yeah something like that

Comment (joemehltretter) [reply vulcanai2/plotters/visualization.py]: @rfratila @nirtiac We can't use `basenetwork.load_model` due to circular dependency issues. I going to remove the load function and push that. Then I'll create an issue regarding this for further discussion.

File: vulcanai2/plotters/visualization.py [hunk 17]
```diff
@@ -7,6 +7,7 @@
 import pickle
 from datetime import datetime
 from .utils import GuidedBackprop, get_notable_indices
+from ..models import basenetwork
```
Comment (rfratila) [inline vulcanai2/plotters/visualization.py]: You should also remove this import to avoid the circular dependency.

File: vulcanai2/plotters/visualization.py (hunk 9 above)
Comment (rfratila) [reply vulcanai2/plotters/visualization.py]: do you still need get_time @joemehltretter ?

File: vulcanai2/plotters/visualization.py (hunk 10 above)
Comment (rfratila) [reply vulcanai2/plotters/visualization.py]: we've created issue #109 to flush this out a little more

File: vulcanai2/tests/plotters/test_visualization.py (hunk 13 above)
Comment (rfratila) [reply vulcanai2/tests/plotters/test_visualization.py]: has this been addressed @joemehltretter ?

File: vulcanai2/tests/plotters/test_visualization.py (hunk 15 above)
Comment (rfratila) [reply vulcanai2/tests/plotters/test_visualization.py]: Have you had a chance to update this @joemehltretter ? just a nitpick thing

**Comment by rfratila** [comment]
@joemehltretter I'm getting the deprecated error from `/Users/robertfratila/miniconda2/envs/py36/lib/python3.6/site-packages/numpy/matrixlib/defmatrix.py:68:` what version of bumpy do you have?

**Comment by joemehltretter** [comment]
@rfratila I am using 1.14.5

File: vulcanai2/tests/plotters/test_visualization.py (hunk 13 above)
Comment (joemehltretter) [reply vulcanai2/tests/plotters/test_visualization.py]: @rfratila Yes, the addition of the `file_created` variable fixed this issue.

Comment (rfratila) [reply vulcanai2/tests/plotters/test_visualization.py]: ok great!

File: vulcanai2/models/basenetwork.py [hunk 18]
```diff
@@ -563,6 +566,9 @@ def fit(self, train_loader, val_loader, epochs,
                 fig_number = plt.gcf().number + 1 if plt.fignum_exists(1) else 1
                 plt.show()
 
+            if save_path:
+                time = get_time()
+                save_path = save_path + '/' + str(self.name) + '_' + time
```
Comment (rfratila) [inline vulcanai2/models/basenetwork.py]: you should use the `get_save_path` no?

**Comment by rfratila** [comment]
@joemehltretter I get the warning with version `1.15.2`

**Comment by joemehltretter** [comment]
@rfratila I updates my numpy and I'm getting the warning. However, the data being sent is `np.ndarray`, which is what it is requesting.

**Comment by rfratila** [comment]
@joemehltretter it seems to be a problem with scipy which is still unresolved. see https://github.com/scikit-learn/scikit-learn/issues/12327

**Comment by joemehltretter** [comment]
@rfratila yeah, that's where I left off on my research last night

File: vulcanai2/models/basenetwork.py (hunk 18 above)
Comment (rfratila) [reply vulcanai2/models/basenetwork.py]: @joemehltretter shouldn't it be something like `save_path = get_save_path(save_path, vis_type=self.name)` ? then we can get rid of `get_time()`

**Comment by rfratila** [comment]
I think we can leave it as is @joemehltretter . If you address the last couple comments we're ready to go.

File: vulcanai2/models/basenetwork.py (hunk 18 above)
Comment (joemehltretter) [reply vulcanai2/models/basenetwork.py]: @rfratila I update this in a recent push. I have it as `vis_type=str(self.name)`. I can remove the `str` if you'd like?

Comment (rfratila) [reply vulcanai2/models/basenetwork.py]: yeah you don't need the `str()` since we check that it's string before when instantiating it.
